src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -1027,76 +943,116 @@
        (when include-hash?
          {:hash (sutils/munge-hash-for-storage
                  (shash/fact-identity-hash fact-data))})))
-    ;; Ensure that all the required paths and values exist, and then
-    ;; insert the new facts.
     (let [paths-and-valuemaps (facts/facts->paths-and-valuemaps values)
           pathstrs (map (comp facts/factpath-to-string first) paths-and-valuemaps)
-          valuemaps (map second paths-and-valuemaps)
-          vhashes (map :value_hash valuemaps)
-          paths-to-ids (realize-paths! pathstrs)
-          vhashes-to-ids (realize-values! valuemaps)
-          pairs (map #(vector (get paths-to-ids %1)
-                              (get vhashes-to-ids %2))
-                     pathstrs vhashes)]
-      (insert-facts-pv-pairs! (certname-to-factset-id certname)
-                              pairs)))))
+          paths->ids (realize-paths! pathstrs)
+          valuemaps (map second paths-and-valuemaps)]
+      (insert-facts! (certname-to-factset-id certname)
+                     (map paths->ids pathstrs)
+                     valuemaps)))))
+
+(defn- update-existing-values [factset-id change-info]
+  (when (seq change-info)
+    (sql/db-do-prepared
+     jdbc/*db* true
+     (cons
+      (str "update facts set"
+           "    value_type_id = ?,"
+           "    value_boolean = ?,"
+           "    value_integer = ?,"
+           "    value_float = ?,"
+           "    value_string = ?,"
+           "    value = ?,"
+           "    large_value_hash = ?"
+           "  where factset_id = ? and fact_path_id = ?")
+      (for [[_ path path-id new-vm] change-info]
+        [(:value_type_id new-vm)
+         (:value_boolean new-vm)
+         (:value_integer new-vm)
+         (:value_float new-vm)
+         (:value_string new-vm)
+         (:value new-vm)
+         (:large_value_hash new-vm)
```

Comment:
Though we should have some tests coming up soon: https://tickets.puppetlabs.com/browse/PDB-3269